hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -684,11 +691,22 @@ private ValidationResult validate(
     private record ValidationResult(
             @NonNull PreHandleResult.Status status, @NonNull ResponseCodeEnum responseCodeEnum) {}
 
+    /**
+     * Rolls back the stack and sets the status of the transaction in case of a failure.
+     * @param rollbackStack whether to rollback the stack. Will be false when the failure is due to a
+     *                      {@link HandleException} that is due to a contract call revert.
+     * @param status the status to set
+     * @param stack the save point stack to rollback
+     * @param recordListBuilder the record list builder to revert
+     */
     private void rollback(
+            final boolean rollbackStack,
             @NonNull final ResponseCodeEnum status,
             @NonNull final SavepointStackImpl stack,
             @NonNull final RecordListBuilder recordListBuilder) {
-        stack.rollbackFullStack();
+        if (rollbackStack) {
+            stack.rollbackFullStack();
+        }
```

Comment:
Using the stack rollback mechanism for everything _but_ gas feels weird. Is there really not a way we can charge gas fees without having to pass a boolean? E.g. some point in the handle workflow, or a common entrypoint for smart contracts

Reply:
Yes we will fix that post-migration to mod-service.

Reply:
It seems all changes look like balance changes in account store for gas charged. So, no other way to differentiate as I understood

Reply:
Ok cool. A `// FUTURE` reminder would be nice (but not required unless tests are already failing 😅)